### DISCUSSION TITLE
antd@5.12.x css变量模式，如何与tailwind集成？

### DISCUSSION THREAD

**Discussion by hemengke1997** [ant-design/ant-design#46527] [Q&A] [answered]
感谢antd团队的付出，antd支持css var模式了！

如何在tailwind中取到这些css变量？这样的话，在代码中可以很方便地使用这些变量

**Comment by MadCcc** [comment, marked as answer]
我不是 tailwindcss 的深度用户，如果可以的集成的话 antd 可以提供 token 到 css 变量的 map 来方便使用。
需要注意的是后续有对 css 变量进行混淆来优化 css 体积的想法，所以不要直接在代码中写死 css 变量比较好。

**Reply by hemengke1997** [reply]
感谢回复。请问现在antd对于css的规划，是否希望用户不写.css文件？因为在css中，无可避免会有hard-code
用户如果希望拥有antd的动态css能力，好像只能在tsx中，通过类似 `useToken` 这种api来获取。我的开发习惯一直是把样式和内容区分开的，所以现在使用antd@5.x会有些别扭。BTW，请问有动态主题这方面比较好的项目示例吗？想了解一下官方团队是如何写样式的

**Reply by MadCcc** [reply]
可以使用 https://ant-design.github.io/antd-style/zh-CN

**Reply by Wxh16144** [reply]
> 需要注意的是后续有对 css 变量进行混淆来优化 css 体积的想法

插一句，我之前也有思考过这个问题，如果我就一个 `normal.css`  我也期望可以直接用 `color: var(--ant-color-primary)` 的形式。

如果用内联样式可能是这样:

```tsx
import { theme } from 'antd';

const Foo = () => {
  const { token } = theme.useToken();

  return <span style={{ color: token.colorPrimary }}>Foo</span>;
};
```

如果用 antd-style 可能是这样：

```tsx
import { createStyles } from 'antd-style';

const useStyle = createStyles(({ css, token }) => ({
  foo: css({
    color: token.colorPrimary,
  }),
}));

const Bar = () => {
  const { styles } = useStyle();
  return <span className={styles.foo}>Bar</span>;
};
```

直接 css variables： (这样其实是最方便的)

```css
.my-span {
  color: var(--ant-color-primary);
}
```

```tsx
import './normal.css';

const Zzz = () => {
  return <span className="my-span">Zzz</span>
}
```

所以，个人感觉如果 css variables 用的是 [SeedToken](https://ant.design/docs/react/customize-theme-cn#seedtoken) 倒是可以考虑，并且压缩混淆的时候 skip 混淆 seedtoken ？

**Reply by hemengke1997** [reply]
我也认为使用css变量（第三种方式）最简单，但是现在 `var(--ant-color-primary)` 都获取不到，放在hash类下面了。🥲

**Reply by Wxh16144** [reply]
> 放在hash类下面了。

讲道理我还没有注意到这个问题 23333~，可以看看大佬们的考量。或许不久将来文档中就说明实践方案了 😄

**Reply by lijunping365** [reply]
@hemengke1997 动态主题功能的项目刚好前两天做了一个，github 地址：https://github.com/lijunping365/Next-Antd-Template

**Reply by MadCcc** [reply]
使用 antd 的容器组件（比如 App）包裹一下，所有子元素就都可以使用 antd 的 css 变量了

**Reply by brandon3343** [reply]
> 都获取不到，放在hash类下

感谢，看到这里知道怎么用了，就是改ConfigProvider配置就行。改了一下，不用引入App组件（因为引入App组件css文件会变大很多）。然后就可以获取到所有的变量了。
```typescript
        <ConfigProvider
          theme={{
            algorithm: themeAlgorithm,
            hashed: false,
            cssVar: {
              key: "css-var-custom",
            },
          }}
          wave={{ disabled: true }}
        >
          <div className="ant css-var-custom">
            <Outlet />
          </div>
        </ConfigProvider>
```

**Comment by promiseLC** [comment]
antd如何和scss做动态主题呢

**Reply by MadCcc** [reply]
scss 是做不了动态主题的，只能预先生成几套 css 文件用来切换

**Reply by promiseLC** [reply]
学到了,就是在scss文件预设几套固定主题,然后修改html 属性,来切换是这样么?  那基于antd做动态主题,采用什么方案比较好呢?

**Comment by susanforme** [comment]
> 学到了,就是在scss文件预设几套固定主题,然后修改html 属性,来切换是这样么? 那基于antd做动态主题,采用什么方案比较好呢?

antd 官方用的token,可以动态修改token来实现换肤

**Comment by youjiaqi421** [comment]
如何在ant-desgin-style 里 使用 tailwindcss 呢  目前看起来并没有一个好的整合方案
